library/alloc/src/collections/btree/borrow.rs

```diff
@@ -0,0 +1,42 @@
+use core::marker::PhantomData;
+use core::ptr::NonNull;
+
+/// A mutable borrow of a type T, that allows intermediately borrowing part of itself,
+/// until waking up the borrow of the entire T.
+pub struct DormantMutRef<'t, T> {
+    ptr: NonNull<T>,
+    _marker: PhantomData<&'t mut T>,
+}
+
+impl<'t, T> DormantMutRef<'t, T> {
+    /// Capture a unique borrow and immediately sublet it to something else,
+    /// that contains one or more unique references to a part of the T object.
```

Comment:
I take it if it's more than one they have to point to disjoint parts? That should be said explicitly, I was confused by how there could be multiple unique references.

But also, `DormantMutRef::new` only lets you take out things once. So how do you get multiple parts? Is the idea that you would return a mutable ref and then "take that apart" later on the client side? I feel like that's just a detail of how the client uses the reference it gets, and it is confusing to put it front-and-center here. But maybe I am misunderstanding the API.

Reply:
I only remember that I went plural here after realizing or imagining that `compute_sub` was extracting multiple references, indeed to disjoint parts I would hope. But now I don't see anything doing that. Those taking multiple references, like `iter_mut` delegating the dirty work to `full_range_search`, do not need this stuff because they don't need any post-processing like adjusting the map's root or length.

That doesn't mean it's nonsensical to have `compute_sub` return multiple references, but definitely not to "a part" as I wrote.